### PR TITLE
crimson/osd: dump each shard seastar metrics

### PR DESCRIPTION
**Pull request by liu-chunmei**
Signed-off-by: chunmei-liu <chunmei.liu@intel.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
why not using prometheus ? otherwise we'd have to expose all metrics using asok.

**Comment by athanatos** [comment]
@tchaikov It's useful for development not to have to stand up a prometheus instance.  I'd actually be perfectly happy with all stats being available via either.

File: src/crimson/admin/osd_admin.cc [hunk 1]
```diff
@@ -226,22 +226,35 @@ class DumpMetricsHook : public AdminSocketHook {
...
-      if (!prefix.empty() && full_name.compare(0, prefix.size(), prefix) != 0) {
-        continue;
-      }
-      for (const auto& [labels, metric] : metric_family) {
-        if (metric && metric->is_enabled()) {
-          dump_metric_value(f.get(), full_name, *metric, labels);
+    f->open_array_section("metrics");
+    return seastar::do_for_each(
+      seastar::smp::all_cpus().begin(),
+      seastar::smp::all_cpus().end(),
```
Comment (athanatos) [inline src/crimson/admin/osd_admin.cc:237]: Let's make this a reusable helper in smp_helpers.h by splitting sharded_map_seq into sharded_map_seq and reactor_map_seq.  Something like:

```

/**
 * reactor_map_seq
 *
 * Invokes f on each reactor sequentially, Caller may assume that
 * f will not be invoked concurrently on multiple cores.
 */
template <typename F>
auto reactor_map_seq(F &&f) {
  using ret_type = decltype(f());
  if constexpr (is_errorated_future_v<ret_type>) {
    auto ret = crimson::do_for_each(
      seastar::smp::all_cpus().begin(),
      seastar::smp::all_cpus().end(),
      [f=std::move(f)](auto core) mutable {
	return seastar::smp::submit_to(
	  core,
	  [&f] {
	    return std::invoke(f);
	  });
      });
    return ret_type(ret);
  } else {
    return seastar::do_for_each(
      seastar::smp::all_cpus().begin(),
      seastar::smp::all_cpus().end(),
      [f=std::move(f)](auto core) mutable {
	return seastar::smp::submit_to(
	  core,
	  [&f] {
	    return std::invoke(f);
	  });
      });
  }
}

/**
 * sharded_map_seq
 *
 * Invokes f on each shard of t sequentially.  Caller may assume that
 * f will not be invoked concurrently on multiple cores.
 */
template <typename T, typename F>
auto sharded_map_seq(T &t, F &&f) {
  return reactor_map_seq(
    [&t, f=std::forward<F>(f)]() mutable {
      return std::invoke(f, t.local());
    });
}
```

**Comment by tchaikov** [comment]
> @tchaikov It's useful for development not to have to stand up a prometheus instance. I'd actually be perfectly happy with all stats being available via either.

but we can just consume the metrics exposed by the builtin prometheus exporter using curl or wget. that's also quite convenient.

**Comment by liu-chunmei** [comment]
> > @tchaikov It's useful for development not to have to stand up a prometheus instance. I'd actually be perfectly happy with all stats being available via either.
> 
> but we can just consume the metrics exposed by the builtin prometheus exporter using curl or wget. that's also quite convenient.

@tchaikov could you please give me some instructions how to use curl or wget to get seastar metrics?

**Comment by athanatos** [comment]
Does crimson currently start the prometheus endpoint by default?  Either way, we can probably merge this for now to unblock @liu-chunmei's work.

**Comment by tchaikov** [comment]
> Does crimson currently start the prometheus endpoint by default?

no, it was disabled by @rzarzynski in 69eeba07267a449c5522a67cb0c31c995bb6908d .

>  Either way, we can probably merge this for now to unblock @liu-chunmei's work.

yup! makes sense.

**Comment by tchaikov** [comment]
> > > @tchaikov It's useful for development not to have to stand up a prometheus instance. I'd actually be perfectly happy with all stats being available via either.
> > 
> > 
> > but we can just consume the metrics exposed by the builtin prometheus exporter using curl or wget. that's also quite convenient.
> 
> @tchaikov could you please give me some instructions how to use curl or wget to get seastar metrics?

i will try to note this down in the section of https://docs.ceph.com/en/latest/dev/crimson/crimson/#prometheus-text-protocol next week.

**Review by cyx1231st — APPROVED** [review]

**Pull request merged**